progress/progress.go

```diff
@@ -0,0 +1,192 @@
+func WithDefaultScaledRamp() Option {
+	return WithScaledRamp("#00dbde", "#fc00ff")
+}
+
+// WithScaledRamp scales the gradient to fit the width of the filled portion of
+// the progress bar.
+func WithScaledRamp(colorA, colorB string) Option {
+	return func(m *Model) error {
+		return m.setRamp(colorA, colorB, true)
+	}
+}
+
+// WithSolidFill sets the progress to use a solid fill with the given color.
+func WithSolidFill(color string) Option {
+	return func(m *Model) error {
+		m.FullColor = color
+		m.useRamp = false
+		return nil
+	}
+}
+
+// WithoutPercentage hides the numeric percentage.
+func WithoutPercentage() Option {
+	return func(m *Model) error {
+		m.ShowPercentage = false
+		return nil
+	}
+}
+
+// WithWidth sets the initial width of the progress bar. Note that you can also
+// set the width via the Width property, which can come in handy if you're
+// waiting for a tea.WindowSizeMsg.
+func WithWidth(w int) Option {
+	return func(m *Model) error {
+		m.Width = w
+		return nil
+	}
+}
+
+// Model stores values we'll use when rendering the progress bar.
+type Model struct {
+
+	// Total width of the progress bar, including percentage, if set.
+	Width int
+
+	// "Filled" sections of the progress bar
+	Full      rune
+	FullColor string
+
+	// "Empty" sections of progress bar
+	Empty      rune
+	EmptyColor string
+
+	// Settings for rendering the numeric percentage
+	ShowPercentage  bool
+	PercentFormat   string // a fmt string for a float
+	PercentageStyle *termenv.Style
+
+	useRamp    bool
+	rampColorA colorful.Color
+	rampColorB colorful.Color
+
+	// When true, we scale the gradient to fit the width of the filled section
+	// of the progress bar. When false, the width of the gradient will be set
+	// to the full width of the progress bar.
+	scaleRamp bool
+}
+
+// NewModel returns a model with default values.
+func NewModel(opts ...Option) (*Model, error) {
+	m := &Model{
+		Width:          defaultWidth,
+		Full:           '█',
+		FullColor:      "#7571F9",
+		Empty:          '░',
+		EmptyColor:     "#606060",
+		ShowPercentage: true,
+		PercentFormat:  " %3.0f%%",
+	}
+
+	for _, opt := range opts {
+		if err := opt(m); err != nil {
+			return nil, err
+		}
+	}
+
+	return m, nil
+}
+
+// View renders the progress bar as a given percentage.
+func (m Model) View(percent float64) string {
+	b := strings.Builder{}
+	if m.ShowPercentage {
+		percentage := fmt.Sprintf(m.PercentFormat, percent*100)
+		if m.PercentageStyle != nil {
+			percentage = m.PercentageStyle.Styled(percentage)
+		}
+		m.bar(&b, percent, ansi.PrintableRuneWidth(percentage))
+		b.WriteString(percentage)
+	} else {
+		m.bar(&b, percent, 0)
+	}
+	return b.String()
+}
+
+func (m Model) bar(b *strings.Builder, percent float64, textWidth int) {
+	var (
+		tw = m.Width - textWidth        // total width
+		fw = int(float64(tw) * percent) // filled width
+		p  float64
+	)
+
+	if m.useRamp {
+		// Gradient fill
+		for i := 0; i < fw; i++ {
+			if m.scaleRamp {
```

Comment:
Looks  like `m.scaleRamp` literally does nothing :smile:

Reply:
The lines are similar looking but not the same:

```go
tw = m.Width - textWidth        // total width
fw = int(float64(tw) * percent) // filled width

p = float64(i) / float64(fw) // first case
p = float64(i) / float64(tw) // second case
```